test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,401 @@
+            }
+
+            Object[][] result = CLUSTER.coordinator(1).execute(select.toString(), specification.readCL);
+
+            if (specification.readCL == QUORUM)
+            {
+                CLUSTER.get(2).runOnInstance(() -> {
+                    Keyspace keyspace = Objects.requireNonNull(Schema.instance.getKeyspaceInstance(KEYSPACE));
+                    SecondaryIndexManager sim = keyspace.getColumnFamilyStore(TEST_TABLE_NAME).indexManager;
+                    for (String s : Arrays.asList("ck_idx", "s_idx", "a_idx", "b_idx"))
+                        sim.makeIndexQueryable(sim.getIndexByName(s), Index.Status.BUILD_SUCCEEDED);
+                });
+
+                SAIUtil.waitForIndexQueryable(CLUSTER, KEYSPACE);
+            }
+
+            return result;
+        }
+    }
+
+    @Parameterized.Parameter
+    public Specification specification;
+
+    @Parameterized.Parameters(name = "{0}")
+    public static List<Object[]> parameters()
+    {
+        List<Object[]> parameters = new ArrayList<>();
+        
+        // Each test scenario operates over a different partition key.
+        int partitionKey = 0;
+
+        for (ConsistencyLevel readCL : new ConsistencyLevel[] { ALL, QUORUM })
+        {
+            for (boolean restrictPartitionKey : new boolean[] { false, true })
+            {
+                for (boolean flushPartials : new boolean[] { false, true })
+                {
+                    for (String[] columns : new String[][] { { "ck", "a" }, { "ck", "s" }, { "s", "a" }, { "a", "b" }, { "a" } })
+                        for (boolean existing : new boolean[] { false, true })
+                            parameters.add(new Object[]{ new Specification(restrictPartitionKey, columns, readCL, existing, StatementType.UPDATE, partitionKey++, flushPartials) });
+
+                    for (String[] columns : new String[][] { { "s", "a" }, { "a", "b" }, { "a" } })
+                        // These scenarios assume existing data.
+                        parameters.add(new Object[]{ new Specification(restrictPartitionKey, columns, readCL, true, StatementType.DELETE, partitionKey++, flushPartials) });
+                }
+            }
+        }
+
+        return parameters;
+    }
+
+    @Test
+    public void testPartialUpdateResolution()
+    {
+        Model model = new Model(specification);
+
+        // Write and repair a version of the row that contains values for all columns we might query:
+        if (specification.existing)
+        {
+            model.writeAndRepairRow();
+            model.validateCurrent();
+        }
+
+        // Introduce a partial update that spans replicas:
+        model.updatePartially();
+
+        if (specification.flushPartials)
+            CLUSTER.stream().forEach(i -> i.flush(KEYSPACE));
+
+        // TODO: Remove, but for now enabling trivially passes all tests:
+        //CLUSTER.get(1).nodetoolResult("repair", KEYSPACE).asserts().success();
```

Comment:
If you disable RFP in `DataResolver` and repair here, every test trivially passes, of course.